File: packages/javascript/lib/format/safeJSONParse.ts
import type { AnyValueObject } from '../_types/global';

export function safeJSONParse(jsonString: any, fallback = {}): AnyValueObject {
	if (!jsonString) {
		return fallback;
	}

	if (Array.isArray(jsonString) || typeof jsonString === 'object') {
		try {
			return JSON.parse(JSON.stringify(jsonString));
		} catch {
			return fallback;
		}
	}

	try {
		return JSON.parse(jsonString);
	} catch {
		return fallback;
	}
}

File: packages/python/qsu/format/safeJSONParse.py
import json
from typing import Any, Optional


def safeJSONParse(jsonString: Any, fallback: Optional[dict] = None) -> dict:
	if fallback is None:
		fallback = {}

	if not jsonString:
		return fallback

	if isinstance(jsonString, (list, dict)):
		try:
			return json.loads(json.dumps(jsonString))
		except (TypeError, ValueError):
			return fallback

	try:
		return json.loads(jsonString)
	except (TypeError, ValueError):
		return fallback
